[GridRow] Introduce GridRow, a columnar layout element

nononoah opened the pull request:
### Summary
This PR introduces `GridRow` which is described below. 

Once a round of discussion topics are resolved and it receives approval, I plan to open a second PR with tests, changelog updates, and so on.

### Achieving consistent, adaptive column layout
`GridRow` is a new `Element` meant to help users write concise, consistent, and adaptive column-based grids.
​
### Background

The `Stack` API is not well-suited to writing `Rows` that layout data in columns. For such layouts, it is essential that sizing is
  - consistent from `Row` to `Row`
  - adaptive across various device sizes

As is, `Stack` layout convenience is oriented around sizing children based on their content size. Notably, we want to avoid that feature in this case, as it is conflict with our first requirement. 
​
Nevertheless, one can achieve the desired effect by wrapping `Rows` in `GeometryReaders`, calculating available layout width, and applying fractions of it to children wrapped in `ConstrainedSize` elements.
​
This is a rather involved, if reasonable, approach. Since this is a common UI paradigm, however, I believe we should provide a simpler path.
​
### Design
`GridRow` layout will support the following:
  - spacing
  -  vertical alignment
  - children with fixed widths
  - children with fractional widths¹

¹Fractional width in this case always means "the fraction of available layout space, after margins and fixed-width children are laid out."
​
Note the absence of horizontal distribution.
​
`GridRow` will exist separately from `Row` and `Stack`, and house its own layout algorithm. The layout algorithm will be relatively simple:
  - determine available layout width by subtracting margins and fixed-width children
  - layout children from left to right, applying the appropriate width and offset
​
### Usage examples
```swift
GridRow { row in
  row.spacing = 8
  row.verticalAlignment = .fill
  row.add(child: .absolute(author, 50))
  row.add(child: .proportional(body, 0.75))
  row.add(child: .proportional(date, 0.25))
}
```

### Demo
![grid-row](https://user-images.githubusercontent.com/1542313/113310899-b60e9e00-92d6-11eb-8e57-63538971a38f.png)

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,191 @@
+///
+/// ## Expected layout:
+///
+/// ┌────────────────────────────┬─┬────────┬─┬──────────────────┐
```
Fwiw dropping this into a ``` block will make it render correctly in Xcode's docs preview

nononoah replied:
Oh really? I actually found the existing doesn't render properly because it's not in a block. If I update it properly, I see the following:

![Screen Shot 2021-04-01 at 2 35 09 PM](https://user-images.githubusercontent.com/1542313/113338764-8cb23a00-92f7-11eb-8ad5-4e32b0900f25.png)

nononoah replied:
Ahh, right! I somehow deleted the top ticks. Thanks for the catch!

nononoah replied:
Fixed.

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,191 @@
+    public var children: [Child] = []
+
+    // MARK: - initialization -
+    public init(configure: (inout GridRow) -> Void) {
```
Nit: `= { _ in }` in case someone doesn't want to configure inline for some reason.

nononoah replied:
Fixed.

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,191 @@
+                sizes[index] = CGSize(width: width, height: height)
+            }
+
+            // Use the tallest child to set the height of all children.
```
Would it make sense to allow a way to align the content top/bottom/centered here?

nononoah replied:
I'm open to it. As is, one would have to wrap children in `Alignment` elements to achieve this effect, which isn't particularly nice.

nononoah replied:
I added this feature in the following commit. The additional layout complexity is actually not too bad:

```swift
// Begin computing frames by reading the vertical alignment. Vertical alignment determines the y-origin
// and, in the `.fill` case, the height of the children.
// The tallest child will be used to size the row.
let rowHeight = sizes.lazy.map { $0.height }.max() ?? 0
let yVector: (CGSize) -> (origin: CGFloat, height: CGFloat)
switch verticalAlignment {
case .fill:
    // Use the tallest child to set the height of all children.
    yVector = { _ in (origin: 0, height: rowHeight) }
case .top:
    yVector = { size in (origin: 0, height: size.height) }
case .center:
    yVector = { size in (origin: (rowHeight - size.height) / 2.0, height: size.height) }
case .bottom:
    yVector = { size in (origin: rowHeight - size.height, height: size.height) }
}
```

Basically, we need to update `yOrigin` and `height` based on alignment. Additionally, we now need to read all frames to determine row height.

I used `lazy` because I hoped it would lead to smaller allocations, but I'm happy to cut it out.

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,191 @@
+
+    // MARK: - GridRow+Element -
+    public var content: ElementContent {
+        ElementContent(layout: GridRowLayout(spacing: spacing)) { builder in
```
If you wanted to, you could also use a `LayoutWriter` here to not need a separate layout – up to you! Just a bit less boilerplate w/ a LayoutWriter.

nononoah replied:
Interesting, let me look into this.

nononoah replied:
I ended up preferring the organization of a separate type f(or mostly aesthetic reasons) given the relative complexity of the layout here. I'm happy to revisit.

n8chur commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,231 @@
+/// │                            │ │        │ │                  │
+/// ●──────────── 75  ───────────● ●── 25 ──● ●─────── 50 ───────●
+/// └────────────────────────────┴─┴────────┴─┴──────────────────┘
+/// ●──────────────────────────── 166 ───────────────────────────●
```
Should this value be 150?

n8chur replied:
It might be helpful to use a value where the proportional parts don't add up to 100 pts to make it more clear how space is distributed.

nononoah replied:
It's subtle, but the extra 16 is attributed to spacing. I'll refine this diagram for clarity.

n8chur replied:
Ah, missed the spacing—makes sense!

nononoah replied:
Cleaned this up a bit by adding numbers to better illuminate spacing and doubling the other figures used.

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+        }
+
+        // MARK: - convenience initialization -
+        public static func absolute(_ element: Element, _ width: CGFloat) -> Child {
```
Nit: I would make the `element` the last parameter here, so the syntax of creating elements inline is cleaner. This is generally the style we try to follow in blueprint. Eg:

```
.absolute(50, Label(text: "...") {
   $0.font = ...
   $0.alignment = ...
})
```

Vs

```
.absolute( Label(text: "...") {
   $0.font = ...
   $0.alignment = ...
}, 50)
```

nononoah replied:
Great point. I'll update, thanks.

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(1)  -> 25 (100 * 1/4)
+        ///     Child B: .proportaional(3) -> 75 (100 * 3/4)
```
proportaional -> proportional

nononoah replied:
:bless:

kyleve commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+        /// ## Example:
+        ///     Layout width: 100
+        ///     Child A: .proportional(0.25)  -> 25 (100 * 1/4)
+        ///     Child B: .proportaional(0.75) -> 75 (100 * 3/4)
```
Same spelling error here

kyleve commented:
(Also, don't forget to update the CHANGELOG.md!)

nononoah commented:
> (Also, don't forget to update the CHANGELOG.md!)

For review clarity, I intended to open a  second PR with CHANGELOG updates, tests, and potentially a demo, then merge them down when we're good to go.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+
+        /// Compute child frames in a given layout space.
+        ///
+        /// First, we measure the size of absolutely-sized  children. Next, we measure the size of
```
```suggestion
        /// First, we measure the size of absolutely-sized children. Next, we measure the size of
```

watt commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            // The tallest child will be used to size the row unless it is a layout pass, in which the row's height
+            // is already known.
+            let rowHeight = isExactConstraint
+                ? constraint.height.maximum
```
Again, this is not going to behave well in unconstrained measurement, and probably won't play nicely if it's nested in other containers either (like if it's in a stack with other content).

In general I would recommend avoiding the `maximum` property except for some special uses, like comparing to a range.

nononoah replied:
I think this should be safe to do in the `.exactly` case, which is what I'm getting at here.

watt commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            isExactConstraint: Bool = false,
+            items: [(traits: Width, content: Measurable)]
+        ) -> [CGRect] {
+            var sizes: [CGSize] = (0..<items.count).map { _ in .zero }
```
`Array(repeating: CGSize.zero, count: items.count)` _might_ be slightly better for this?

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            }
+
+            // Measure the the row to fit its tallest child.
+            let height = _frames(in: constraint, items: items).lazy.map { $0.height }.max() ?? 0
```
Correct, that is the sole advantage.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+        ///   - items: Children to layout.
+        private func _frames(
+            in constraint: SizeConstraint,
+            isExactConstraint: Bool = false,
```
This is in spirit identical, but meant to avoid the additional type surface area, as it seems to be the only additional bit of information worth encoding.

That said, perhaps I should just use the existing type. Let me look into it.

nononoah replied:
To make full use of `VectorConstraint`, I'd have to open it (and a few related extensions) up a bit. After exploring it, I decided to stick to our own type, which is scoped down to exactly the concerns we have. This type, of course, is nothing more than a `SizeConstraint` and a `Bool` at heart, but it does make the layout code marginally more legible.

Please let me know what you think!

nononoah replied:
After playing around a bit with a few approaches, my impression is that this adds a fair bit of mostly unused complexity to `GridRow`.

That said, if we assume the average reader to already be familiar with Row, it's perhaps not a significant burden to add.

i.e.

```swift
- SizeConstraint(width: .atMost(width), height: constraint.height)
+ SizeConstraint(width: .atMost(width), height: constraint.cross.sizeConstraint)
```

```swift
- let width = content.measure(in: constraint).width
+ let width = content.measure(in: constraint.constraint(axis: .horizontal)).width
```

```swift
- let rowHeight = isExactConstraint 
-      ? constraint.height.maximum 
-      : sizes.lazy.map { $0.height }.max() ?? 0
+ let rowHeight: CGFloat
+ switch constraint.cross {
+ case .exactly(height): 
+     rowHeight = height
+ case .atMost, .unconstrained:
+    rowHeight = sizes.lazy.map { $0.height }.max() ?? 0
}
```

watt replied:
OK, thanks for considering. Sometimes I find that modeling that way can make things easier to understand (see my misunderstanding of the `rowHeight` ternary below), but if it's less cumbersome to write this way, that's fine! We can ensure nothing unexpected happens with adequate test coverage.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            let availableWidth = constraint.width.maximum - (spacing * CGFloat(items.count - 1)) - absoluteItemWidth
+            // Determine the number of portions and the width of each portion.
+            let portionCount = proportionallySized.lazy.map { $0.proportion }.reduce(0, +)
+            let portion = availableWidth / portionCount
```
Great catch, thought it really would be quite odd to use this type without at least one `.proportional` child. 

I'll make sure this is fixed after I update the layout approach.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            }
+
+            // Measure proportionally-sized children, allotting them a portion of the available space.
+            let availableWidth = constraint.width.maximum - (spacing * CGFloat(items.count - 1)) - absoluteItemWidth
```
This makes a lot of sense to me, though similarly to what follows below, we want to make sure we account for the `.exactly` case, where it is true that we can use the given value.

Because I was only testing in vertical containers (`ScrollView` and `Column` mostly), I missed the potential flaws in this shortcut.

nononoah replied:
I followed this approach, and ended up with the following:

```
var scale: CGFloat = 0
items.forEach { (_, proportion, content) in
    let width = content.measure(in: constraint).width
    scale = max(scale, width / proportion)
}

items.forEach { (index, proportion, content) in
    let width = scale * proportion
    let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
    sizes[index] = CGSize(width: width, height: content.measure(in: constraint).height)
}
```

I'll work on making this as legible as possible. Thanks for the guidance.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+            // Measure the the row to fit its tallest child.
+            let height = _frames(in: constraint, items: items).lazy.map { $0.height }.max() ?? 0
+
+            return CGSize(width: constraint.width.maximum, height: height)
```
Got it. I'll update this to reduce frames into a given size, similar to how `Stack` reduces vectors into the same.

nononoah commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,244 @@
+
+// MARK: - child modeling -
+public extension GridRow {
+    enum VerticalAlignment: Equatable {
```
I'll look into reuse here. Thanks for the tip.

nononoah replied:
While I didn't end up reusing the vertical alignment code, I was able to conserve this type without much churn.

nononoah commented:
@watt, I took a shot at addressing all of your concerns here, and left comments where I didn't quite follow your lead. I'd appreciate another look when you have time.

n8chur commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    // MARK: - mutations -
+    public mutating func add(child: Child) {
```
Thought: I think this is probably fine (or even perhaps preferable if we were consistent about the approach), but should the values that make up `Child` be provided as separate parameters here? Maybe just as a separate convenience func?

Alternatively, should we migrate `Stack` to use this style?

No need to take action, but it might be worth driving towards a consistent API overall if there's no technical reason to stray.

nononoah replied:
It's a great question. The main difference, I think, is that `GridRow` uses a `struct`, while `Row` uses a tuple, which makes the spelling of these sites a bit awkward. What's more, `Row` can provide reasonable defaults for `grow` and `shrink` priority, but I can't.

I will replace the existing sugar with a style more in keeping with `Row`. Does something like 

```swift
public mutating func addAbsolute(child: Element, width: CGFloat) { ... }
public mutating func addProportional(child: Element, proportion: CGFloat) { ... }
```

roughly match your expectations?

n8chur replied:
Ah, the lack of reasonable defaults is an interesting point.

I was thinking:
```swift
public mutating func add(width: Width, key: AnyHashable? = nil, child: Element) {
    add(child: .init(width: width, key: key, element: child)
}
```
but what you have above seems fine as well.

Curious what the other Blueprint folks think, but it's probably fine to land as is and provide additional conveniences if they seem helpful 🤷

watt replied:
> ```swift
> public mutating func add(width: Width, key: AnyHashable? = nil, child: Element) {
>     add(child: .init(width: width, key: key, element: child)
> }
> ```
> 
> Curious what the other Blueprint folks think, but it's probably fine to land as is and provide additional conveniences if they seem helpful 🤷

This is probably the closer to the API I would build personally, but I don't have strong feelings about it. Both of them can support new params (e.g. alignment guides) or convenience methods.

nononoah replied:
Thanks for the suggestion. I'll update before merging.

watt commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,303 @@
+            case .align(let id) where id == .bottom:
+                return { height in (rowHeight - height,height) }
+            case .align:
+                fatalError("GridRow supports fill, top, center, and bottom alignment.")
```
Hmm, I guess a follow-up for me would be to extract the cross-axis alignment logic from stacks so this is easier to implement.

watt commented on BlueprintUI/Sources/Layout/GridRow.swift:
```diff
@@ -0,0 +1,303 @@
+            return items.map { (index, proportion, content) in
+                let width = scale * proportion
+                let constraint = SizeConstraint(width: .atMost(width), height: constraint.height)
+                let size = CGSize(width: width, height: content.measure(in: constraint).height)
```
Can we skip this second measurement and use the height from the first?

nononoah replied:
The overloading of the `constraint` here makes this a bit less legible, so I intend to fix it.

The first measurement measures the element against an unconstrained constraint. The second measurement occurs against a width-constrained constrained.

Because the height of some elements might grow when the width is constrained, I think we need this second measurement. Please correct me if I've misunderstood!

watt replied:
> The first measurement occurs against an unconstrained constraint. The second measurement occurs against a width-constrained constrained.
> 
> Because the height of some elements might grow when the width is constrained, I think we need this second measurement. Please correct me if I've misunderstood!

If I'm understanding right, the width constraint used for the second measurement is guaranteed to be at least as wide as the size that the element reported in the first measurement. So while an element _could_ report a different size under that constraint, I think it would be unusual. Do you have a use case in mind where this would return a different size? I'm not sure it's a behavior we need to support.

(Having just done a performance pass I'm a bit wary of measurement calls.)

nononoah replied:
The use case I had in mind was a label. 

Given an unconstrained width (the first measurement), it would have a height as tall as a single line. 

Given a constrained width (the second measurement), the height will in some cases be taller.

> the width constraint used for the second measurement is guaranteed to be at least as wide as the size that the element reported in the first measurement

I think that is correct, but I'm not sure it accounts for the height concern.

I also want to avoid unnecessary measurement and did try to reason through each call to see if it could be skipped. Still, I'm a novice, and my naivety is worth testing. Please let me know if you think my case is flawed.

kyleve replied:
This seems correct – if the width deviates at all from previous measurements, you will need to re-measure, because things like labels will re-flow.

watt replied:
OK, here's an example with a label that might help.

```swift
let label = UILabel()
label.text = "Hello world"
```

Let's measure the label with no constraint. Here we're asking the label: given an infinite amount of space, what is the size you want to be?

```swift
let unconstrainedSize = label.sizeThatFits(.zero)
print(unconstrainedSize) // (85.5, 20.5)
```

Now let's take that size, and use it as a constraint:

```swift
let constrainedSize = label.sizeThatFits(unconstrainedSize)
print(constrainedSize) // (85.5, 20.5)
```

The label reports the same size. This is because the constraint is large enough that the label does not need to reflow. We know it is large enough, because the label has already told us that this is the size it wants to be. For any well-behaved element, this should hold true.

nononoah replied:
Got it. Apologies for being a bit dense here, this makes sense. I'll open a PR to fix it.
